File: src/dijkstra/dijkstra.cpp
#include <vector>
#include <string>
#include <fstream>
#include <iostream>
#include <limits>
#include <chrono>

// class for nodes
class Node
{

public:

    Node(int id, std::string path, int length);

    int id;
    std::string path;
    int length;
};

// Constructor
Node::Node(int id, std::string path, int length)
:
    id(id),
    path(path),
    length(length)
{}

// Function that read the topology
int** readTopologie(std::string name, int& nbNode)
{
    std::ifstream ifs(name);
    if(!ifs)
    {
        std::cerr << "file " << name << " not found !" << '\n';
        exit(1);
    }

    ifs >> nbNode;
    int** topology = new int*[nbNode];
    for(int i = 0; i < nbNode; i++)
    {
        topology[i] = new int[nbNode];
        for(int j = 0; j < nbNode; j++)
        {
            ifs >> topology[i][j];
        }
    }
    ifs.close();
    return topology;
}

int main(int argc, char* argv[])
{
    if(argc != 3)
    {
        std::cerr << "2 arguments expected : fileName and iterations" << '\n';
        exit(1);
    }
    int nbIter = std::stoi(argv[2]);
    int nbNode;
    int** topology = readTopologie(argv[1], nbNode);

    // target
    int idNode = 0;

    auto start = std::chrono::steady_clock::now();

    for(int iter = 0; iter < nbIter; iter++)
    {
        std::vector<Node*> nodeDone;
        std::vector<Node*> nodeLeft;

        for(int i = 0; i < nbNode; i++)
        {
            if(i == idNode) // target
                nodeDone.push_back(new Node(i, std::to_string(idNode), 0));
            else if(topology[i][idNode] != 0) // if linked
                nodeLeft.push_back(new Node(i, std::to_string(idNode), topology[i][idNode]));
            else // if not linked
                nodeLeft.push_back(new Node(i, std::to_string(idNode), std::numeric_limits<int>::max()));
        }

        while(!nodeLeft.empty())
        {
            int nodeMinPos = 0;
            for(size_t i = 1, max = nodeLeft.size(); i < max; i++) // looking for the nearest node
            {
                if(nodeLeft[nodeMinPos]->length > nodeLeft[i]->length)
                    nodeMinPos = i;
            }

            Node* minNode = nodeLeft[nodeMinPos];
            minNode->path += " -> " + std::to_string(minNode->id);

            for(Node* n : nodeLeft) // update other nodes
            {
                int dist = topology[minNode->id][n->id];
                if(dist != 0)
                {
                    if(n->length > minNode->length + dist)
                    {
                        n->length = minNode->length + dist;
                        n->path = minNode->path;
                    }
                }
            }
            nodeDone.push_back(minNode);
            nodeLeft.erase(nodeLeft.begin() + nodeMinPos);
        }

        // print result
        /*
        for(size_t i = 0; i < nodeDone.size(); i++)
        {
            std::cout << nodeDone[i]->path << " | " << nodeDone[i]->length << " | " << nodeDone[i]->path << '\n';
        }
        */

        // free memory
        for(int i = 0; i < nbNode; i++)
        {
            delete nodeDone[i];
        }
    }

    auto end = std::chrono::steady_clock::now();
    std::cout << std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count() / nbIter;

    for(int i = 0; i < nbNode; i++)
    {
        delete[] topology[i];
    }
    delete[] topology;

    return 0;
}

File: src/dijkstra/dijkstra.py
#!/usr/bin/env python3.8

import sys
from time import perf_counter_ns

if(len(sys.argv) != 3):

    sys.stderr.write("2 arguments expected : fileName and iterations\n")
    exit(1)

nb_iter = int(sys.argv[2])

# read file
topo = open(sys.argv[1], "r").read().split("\n")
nb_node = int(topo.pop(0))

for i in range(0, nb_node):
    lign = topo[i].split(" ")
    topo[i] = list(map(int, lign))

# target node
id_node = 0

start = perf_counter_ns()

for iter in range(0, nb_iter):
    nodes_done = []
    nodes_left = []

    for i in range(0, nb_node): # initialisation
        node = dict()

        if(i == id_node): # target
            node["id"] = i
            node["path"] = str(id_node)
            node["length"] = 0
            nodes_done.append(node)
        else:
            node["id"] = i
            if(topo[id_node][i] != 0): # if linked
                node["length"] = topo[id_node][i]
                node["path"] = str(id_node)
            else: # if not linked
                node["length"] = float("inf")

            nodes_left.append(node)

    while nodes_left:

        node_min_pos = 0
        for i in range(1, len(nodes_left)): # looking for the nearest node
            if(nodes_left[node_min_pos]["length"] > nodes_left[i]["length"]):
                node_min_pos = i

        node_min = nodes_left[node_min_pos]
        node_min["path"] += " -> " + str(node_min["id"])

        for n in nodes_left: # update other nodes

            dist = topo[node_min["id"]][n["id"]]

            if(dist != 0):

                if(n["length"] > node_min["length"] + dist):

                    n["length"] = node_min["length"] + dist
                    n["path"] = node_min["path"]

        nodes_done.append(nodes_left.pop(node_min_pos))

    # print result
    #for n in nodes_done:
    #    print(n)
end = perf_counter_ns()
print((end - start) // nb_iter, end = "")
